text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,682 @@
+  one of the variations), for extremely high performance random
+  numbers,
+- `ChaChaRng`, a [ChaCha] random number generator, for "secure"
+  user-space random numbers.
+
+[xorshift]: http://en.wikipedia.org/wiki/Xorshift
+
+### Algorithms
+
+#### `OsRng`
+
+An *unbuffered* wrapper around the best available operating system
+primitive, e.g.
+
+- `CryptGenRandom` on Windows,
+- `getrandom(2)` on recent Linux versions,
+- `/dev/urandom` on OSX, BSD, and older linux versions.
+
+(See [the current `std::rand`][current] docs for more info and
+discussion about this, particularly `/dev/urandom`. NB. to future
+readers, the information will still be available in the docs after
+this RFC lands, but may have moved.)
+
+[current]: http://doc.rust-lang.org/nightly/std/rand/#cryptographic-security
+
+The lack of buffering by default means every call will interact with
+the operating system, making it very slow; but this is the only way
+that the RFC author is comfortable to call it cryptographically
+secure, adding complications is risky e.g. naively buffering would not
+be cryptographically secure due to `fork` sharing memory.
+
+This kind of functionality is sometimes called "random device", as a
+possible alternative name.
+
+#### PRNGs
+
+Only two user-space pseudo-random number generators are provided:
+[ChaCha] and [Xorshift][xorshift].
+
+This specifically means the ISAAC and ISAAC64 implementations will be
+moved out of `std::rand`. Compared to ISAAC*, the ChaCha RNG is much
+simpler, has had deeper cryptanalysis and is designed with SIMD
+optimisations in mind so is more suited to being part of a standard
+library.
+
+These two types can be deterministically seeded and would expose this
+functionality as inherent methods (not trait implementations), and
+would implement `Random` to "optimally" randomly seed themselves from
+another source of randomness.
+
+For example, the stable public interface for `ChaChaRng` (and
+`XorShiftRng`) will look like:
+
+```rust
+pub struct ChaChaRng { ... }
+
+impl ChaChaRng {
+    pub fn from_seed(seed: &[u32]) -> ChaChaRng { ... } // seed: [u32; 4] for XorshiftRng
+}
+
+impl Random<FullRange> for ChaChaRng {
+    ...
+}
+impl Reader for ChaChaRng {
+    ...
+}
+impl Rng for ChaChaRng {
+    ...
+}
+```
+
+These names are chosen to be somewhat vague: `ChaChaRng` is actually,
+currently, an implementation of the ChaCha20 algorithm, but the "20"
+value could be able to be customised backwards-compatibly in future,
+with default type parameters. (Similarly Xorshift is strictly
+Xorshift128, but has parameters that could be tweaked.) E.g. in
+hypothetical syntax
+
+```rust
+struct ChaChaRng<static N: u32 = 20> {
+    /* state */
+}
+let default: ChaChaRng = ...;
+let faster: ChaChaRng<8> = ...;
+```
+
+
+#### Alternatives
+
+As per the Alternatives section of the [free functions] discussion,
+these could be named `SecureRng`, `BalancedRng` and `WeakRng` (or have
```

Comment:
http://www.openbsd.org/cgi-bin/man.cgi/OpenBSD-current/man2/getentropy.2

> getentropy is not intended for regular code; please use the arc4random(3) family of functions instead.

You'll be violating the advice given by the OpenBSD documentation by using this cargo cult implementation.